File: go/brainfuck.go
package main

import (
	"brainfuck/interpreter"
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	if len(os.Args) < 2 {
		log.Fatal(fmt.Sprintf("Usage: %s <FILENAME>", os.Args[0]))
		os.Exit(1)
	}

	fileBytes, e := os.ReadFile(os.Args[1])
	if e != nil {
		log.Fatal(fmt.Sprintf("Cannot open file %s: %s", os.Args[0], e))
		os.Exit(1)
	}

	program := string(fileBytes)

	interpreter := interpreter.BrainfuckInterpreter{
		Program:   program,
		JumpTable: nil,
		TapeSize:  30000,
		Cin:       default_cin,
		Cout:      default_cout}
	e = interpreter.BuildJumpTable()
	if e != nil {
		log.Fatal(e)
		os.Exit(1)
	}

	interpreter.Run()

	os.Exit(0)
}

func default_cin() byte {
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return input[0]
}

func default_cout(b byte) {
	fmt.Printf("%s", string(b))
}

File: pascal/Brainfuck.pas
program Brainfuck;
uses BrainfuckInterpreter, BrainfuckUtility;

var
    Debug           : boolean;
    Interpreter     : TBrainfuckInterpreter;

procedure DefaultCout(x: byte);
begin
   Write(Chr(x));
end;

function DefaultCin(): byte;
var ch: char;
begin
   Read(ch);
   DefaultCin := Ord(ch);
end;

function ReadCode(Filename: string): AnsiString;
var
    InputFile       : file of char;
    CurrentError    : integer;
    CurrentErrorStr : string;
    CurrentChar     : char;
begin
    ReadCode := '';

    Assign(InputFile, Filename);
    CurrentError := IOResult;
    if CurrentError <> 0 then begin
        Str(CurrentError, CurrentErrorStr);
        Panic('Cannot open file "' + Filename +'" (code: ' + CurrentErrorStr +')');
    end;

    Reset(InputFile);
    CurrentError := IOResult;
    if CurrentError <> 0 then begin
        Str(CurrentError, CurrentErrorStr);
        Panic('Cannot open file "' + Filename +'" (code: ' + CurrentErrorStr +')');
    end;

    while not Eof(InputFile) do begin
        Read(InputFile, CurrentChar);
        CurrentError := IOResult;
        if CurrentError <> 0 then begin
            Str(CurrentError, CurrentErrorStr);
            Panic('Cannot read file "' + Filename +'" (code: ' + CurrentErrorStr +')');
        end;

        ReadCode := ReadCode + CurrentChar;
    end;

    Close(InputFile);
    if CurrentError <> 0 then begin
        Str(CurrentError, CurrentErrorStr);
        Panic('Cannot clode file "' + Filename +'" (code: ' + CurrentErrorStr +')');
    end;
end;

begin
    if (paramCount() <> 1) and (paramCount() <> 2) then begin
        writeLn('Usage: ', paramStr(0), ' <filename> [-debug]');
        Panic('')
    end;

    Debug := (paramCount() = 2) and (paramStr(2) = '-debug');
    InitializeBrainfuckInterpreter(Interpreter, ReadCode(paramStr(1)), @DefaultCin, @DefaultCout, Debug);
    RunBrainfuckInterpreter(Interpreter, Debug);
end.